File: backend/app/schemas/listing.py
from decimal import Decimal
from typing import Optional, Any, Literal
from uuid import UUID
from datetime import datetime
from pydantic import BaseModel, field_validator


class ListingCreate(BaseModel):
    sku_external_id: Optional[str] = None
    sku_description: str
    sku_brand: str
    price: Decimal
    stock_quantity: int
    condition: str
    listing_type_id: str = "gold_special"

    @field_validator("condition")
    @classmethod
    def condition_must_be_valid(cls, v: str) -> str:
        if v not in ("new", "used"):
            raise ValueError("Condição deve ser 'new' ou 'used'")
        return v

    @field_validator("listing_type_id")
    @classmethod
    def listing_type_must_be_valid(cls, v: str) -> str:
        if v not in ("gold_special", "gold_pro"):
            raise ValueError("Tipo de anúncio deve ser 'gold_special' (Clássico) ou 'gold_pro' (Premium)")
        return v

    @field_validator("price")
    @classmethod
    def price_must_be_positive(cls, v: Decimal) -> Decimal:
        if v <= 0:
            raise ValueError("Preço deve ser maior que zero")
        return v

    @field_validator("stock_quantity")
    @classmethod
    def stock_must_be_positive(cls, v: int) -> int:
        if v < 1:
            raise ValueError("Estoque deve ser pelo menos 1")
        return v


class ListingSummary(BaseModel):
    id: UUID
    product_id: Optional[UUID] = None
    sku_external_id: Optional[str]
    sku_brand: str
    selected_title: Optional[str]
    status: str
    failed_step: Optional[str] = None
    created_via: str
    mlb_id: Optional[str]
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


class ListingPage(BaseModel):
    items: list[ListingSummary]
    total: int
    page: int
    page_size: int


class TitleOption(BaseModel):
    id: UUID
    title_text: str
    ai_score: Optional[Decimal]
    selected: bool

    model_config = {"from_attributes": True}


class AttributeOut(BaseModel):
    id: UUID
    attribute_id: str
    attribute_name: str
    attribute_type: str
    is_required: bool
    value_id: Optional[str]
    value_name: Optional[str]
    source: str
    allowed_values: Optional[list[Any]]

    model_config = {"from_attributes": True}


class JobOut(BaseModel):
    id: UUID
    job_type: str
    celery_task_id: Optional[str]
    status: str
    error_message: Optional[str]
    attempts: int
    started_at: Optional[datetime]
    completed_at: Optional[datetime]
    created_at: datetime

    model_config = {"from_attributes": True}


class ImageOut(BaseModel):
    id: UUID
    url_r2: Optional[str]
    ml_picture_id: Optional[str]
    status: str
    approved: bool
    sort_order: int

    model_config = {"from_attributes": True}


class ImageApproveRequest(BaseModel):
    approved_ids: list[UUID]
    # Tempo (segundos) que um humano levou comparando a imagem gerada por IA
    # contra o dado real antes de aprovar. Opcional: ausente -> fica NULL,
    # comportamento identico ao de antes deste campo existir.
    review_seconds: Optional[int] = None


class ImageEngineConfirmRequest(BaseModel):
    action: Literal["use_gemini", "retry_openai"]


class ListingDetail(ListingSummary):
    sku_description: str
    price: Decimal
    stock_quantity: int
    condition: str
    listing_type_id: str
    ml_category_id: Optional[str]
    error_message: Optional[str]
    description_html: Optional[str] = None
    titles: list[TitleOption] = []
    attributes: list[AttributeOut] = []
    images: list[ImageOut] = []
    jobs: list[JobOut] = []

File: frontend/src/types/listing.ts
export type Condition = "new" | "used"

export type ListingStatus =
  | "draft"
  | "generating_title"
  | "pending_title_approval"
  | "predicting_category"
  | "pending_seller_attributes"
  | "pending_description"
  | "generating_images"
  | "pending_image_approval"
  | "generating_description"
  | "ready_to_publish"
  | "publishing"
  | "published"
  | "published_paused"
  | "failed"

export interface ListingSummary {
  id: string
  sku_external_id: string | null
  sku_brand: string
  selected_title: string | null
  status: ListingStatus
  created_via: "manual" | "batch"
  mlb_id: string | null
  created_at: string
  updated_at: string
  failed_step?: string | null
}

export interface TitleOption {
  id: string
  title_text: string
  ai_score: number | null
  selected: boolean
}

export interface AttributeOut {
  id: string
  attribute_id: string
  attribute_name: string
  attribute_type: string
  is_required: boolean
  value_id: string | null
  value_name: string | null
  source: string
  allowed_values: { id: string; name: string }[] | null
}

export interface ImageOut {
  id: string
  url_r2: string | null
  ml_picture_id: string | null
  status: string
  approved: boolean
  sort_order: number
}

export interface JobOut {
  id: string
  job_type: string
  status: string
  error_message: string | null
  attempts: number
  created_at: string
}

export interface ListingDetail extends ListingSummary {
  sku_description: string
  price: number
  stock_quantity: number
  condition: Condition
  listing_type_id: string
  ml_category_id: string | null
  error_message: string | null
  description_html: string | null
  titles: TitleOption[]
  attributes: AttributeOut[]
  images: ImageOut[]
  jobs: JobOut[]
}

export interface ListingsResponse {
  items: ListingSummary[]
  total: number
  page: number
  page_size: number
}

export const STATUS_LABELS: Record<ListingStatus, string> = {
  draft: "Rascunho",
  generating_title: "Gerando título",
  pending_title_approval: "Aguardando aprovação de título",
  predicting_category: "Predizendo categoria",
  pending_seller_attributes: "Aguardando atributos",
  pending_description: "Aguardando descrição",
  generating_images: "Gerando imagens",
  pending_image_approval: "Aguardando aprovação de imagens",
  generating_description: "Gerando descrição",
  ready_to_publish: "Pronto para publicar",
  publishing: "Publicando",
  published: "Publicado",
  published_paused: "Pausado",
  failed: "Com erro",
}

export const PROCESSING_STATUSES: ListingStatus[] = [
  "generating_title",
  "predicting_category",
  "generating_images",
  "generating_description",
  "publishing",
]

export const WAITING_STATUSES: ListingStatus[] = [
  "pending_title_approval",
  "pending_seller_attributes",
  "pending_image_approval",
  "ready_to_publish",
]

export interface BulkItemResult {
  listing_id: string
  success: boolean
  error?: string | null
}

export interface BulkResult {
  processed: number
  failed: number
  results: BulkItemResult[]
}

export interface AttributeItem {
  attribute_id: string
  attribute_name: string
  value_name: string | null
  value_id: string | null
  is_required: boolean
}

export interface ListingAttributesRow {
  listing_id: string
  sku_external_id: string
  selected_title: string | null
  ml_category_id: string | null
  status: string
  attributes: AttributeItem[]
}
